### PR TITLE
[#10928] Add script to list active instructors within a period

### LINKED ISSUE #10928: Add a script for the admin to generate a list of active instructors
context: We need to inform instructors about the replacement of 'submission confirmation email' with a 'download proof of submission' feature. Otherwise, instructors can end up giving wrong instructions to students.

suggestion: Add a script for the admin to generate a list of emails of instructors who have been active in a certain period in the past.

Example of another admin script `java/teammates/client/scripts/SingleAccountGoogleIdMigrationScript.java`

### PR DESCRIPTION
**Pull request by daongochieu2810**
Fix #10928
Solution outline:
- Filter feedback sessions based on given start and end times
- Get creator email from filtered sessions and print them

### CODE REVIEW

File: src/client/java/teammates/client/scripts/ListActiveInstructors.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * Script to generate emails of active instructors within a period.
+ */
+public class ListActiveInstructors extends RemoteApiClient {
+    @Override
+    protected void doOperation() {
+        //2020/01/01
+        long startTime = 1577808000000L;
+        //2021/02/01
+        long endTime = 1612108800000L;
```
Comment (damithc) [inline src/client/java/teammates/client/scripts/ListActiveInstructors.java:18]: I'll have to update these manually. So, use a human editable format for the dates :-)

File: src/client/java/teammates/client/scripts/ListActiveInstructors.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        int endDate = 1;
+        int endMonth = 2;
+        int endYear = 2021;
+
+        LocalDate startPoint = LocalDate.of(startYear, startMonth, startDate);
+        LocalDate endPoint = LocalDate.of(endYear, endMonth, endDate);
+        long startTime = startPoint.atStartOfDay().toInstant(ZoneOffset.UTC).toEpochMilli();
+        long endTime = endPoint.atStartOfDay().toInstant(ZoneOffset.UTC).toEpochMilli();
+
+        ArrayList<String> activeInstructorEmails = new ArrayList<>();
```
Comment (Derek-Hardy) [inline src/client/java/teammates/client/scripts/ListActiveInstructors.java:31]: I think `Set` would be a more appropriate data structure for `contains` and `add` operations?

**Review by Derek-Hardy — APPROVED** [review]
LGTM

**Comment by damithc** [comment]
I gave it a try. Took 22 minutes to finish, and produced 1.6k results. So, the approach is not very scalable.  It might even time out when we have even more sessions. We have 66K sessions by the way.
If the sessions are given to us in the order they were created, perhaps we can terminate after we hit a session that's outside of the period?

**Comment by daongochieu2810** [comment]
> I gave it a try. Took 22 minutes to finish, and produced 1.6k results. So, the approach is not very scalable. It might even time out when we have even more sessions. We have 66K sessions by the way.
> If the sessions are given to us in the order they were created, perhaps we can terminate after we hit a session that's outside of the period?

got it, I will try to prune it as you suggested

**Comment by rrtheonlyone** [comment]
You do not need to binary search after fetching ALL sessions. 

I believe you can do this with just one query that can filter by created time - like this (just an arbitrary example):

```java
ofy().load().type(CourseStudent.class)
                   .filter("createdAt >", queryEntitiesFrom)
                   .filter("createdAt <=", queryEntitiesTo)
```                    

Furthermore, since the `creatorEmail` is indexed, you can then do a projection on the query to retrieve it - `project()`.

It will look something like - `ofy.load().type(..).filter(..).filter(..).project(..)`

This will allow you to do just one query to fetch everything you need. This should be much faster as the database will help optimize for you.

File: src/client/java/teammates/client/scripts/ListActiveInstructors.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        LocalDate startPoint = LocalDate.of(startYear, startMonth, startDate);
+        LocalDate endPoint = LocalDate.of(endYear, endMonth, endDate);
+        long startTimeInMilli = startPoint.atStartOfDay().toInstant(ZoneOffset.UTC).toEpochMilli();
+        long endTimeInMilli = endPoint.atStartOfDay().toInstant(ZoneOffset.UTC).toEpochMilli();
+        Instant startTime = Instant.ofEpochMilli(startTimeInMilli);
+        Instant endTime = Instant.ofEpochMilli(endTimeInMilli);
+
+        Set<String> activeInstructorEmails = new HashSet<>();
+
+        ofy().load().type(FeedbackSession.class)
```
Comment (rrtheonlyone) [inline src/client/java/teammates/client/scripts/ListActiveInstructors.java:36]: almost there, you have to use a field given in the class as a filter (and the field has to be indexed)

```java
ofy().load().type(FeedbackSession.class)
        .filter("startTime >=", startTime)
        .filter("startTime <=", endTime).project().forEach(feedbackSession -> {
            activeInstructorEmails.add(feedbackSession.getCreatorEmail());
        })
```

Let's use `startTime`

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Minor Change - do test the script locally before pushing in future :)

**Review by rrtheonlyone — APPROVED** [review]
Looks good - prof @damithc do see if this is faster when you have time.

This script finds the list of active instructors using the start time of the feedback sessions they have created. This is just a heuristic - I am not sure if we can find all active instructors using this.

**Comment by damithc** [comment]
> Looks good - prof @damithc do see if this is faster when you have time.

This one takes about 4 minutes. Nice!

**Pull request merged**